src/server.py

```diff
@@ -114,34 +160,40 @@ def delay(self):
     def sound_location(self):
         return self.metadata["sound-file"]
 
-    def _add_fade_in(self, element, time_ms, volume):
-        if not self._fade_control.set(0, 0):
+    def _add_keyframe_pair(self, element, control, time_start, volume_start,
```

Comment:
> I would treat add_keyframe_pair and add_keyframe as a bit "lower-level" functions where the units should always be in GStreamer units, which is nanoseconds.
I am addresing this removing the unit argument and using the default unit by default.

> I would also move the logic from add_fade_out to here, so that we never add a keyframe pair that extends over a loop boundary. 

Do you refer to the case when the duration of the sound is T seconds, and I call the UpdateProperties method at T/2 seconds with a duration of T seconds?

Reply:
Yeah, I'm pretty sure @ptomato wants this case [1] to be made more general than only applying to the volume fade out. (I agree.)

[1] https://github.com/endlessm/hack-sound-server/blob/3b8043b76370af70eae8f9f7615048757fb51744/src/server.py#L174